src/marinade-state/marinade-state.ts

```diff
@@ -133,6 +139,74 @@ export class MarinadeState {
     )
   }
 
+  async getStakeStates(): Promise<StakeState[]> {
+    const stakeAccountInfos = await this.anchorProvider.connection.getProgramAccounts(STAKE_PROGRAM_ID, {
+      filters: [
+        { dataSize: 200 },
+        {
+          memcmp: {
+            offset: 44,
+            bytes: this.marinade.config.stakeWithdrawAuthPDA.toString(),
+          },
+        },
+      ],
+    })
+
+    return stakeAccountInfos.map((stakeAccountInfo) => {
+      const { data } = stakeAccountInfo.account
+      // The data's first 4 bytes are: u8 0x0 0x0 0x0 but borsh uses only the first byte to find the enum's value index.
+      // The next 3 bytes are unused and we need to get rid of them (or somehow fix the BORSH schema?)
+      const adjustedData = Buffer.concat([
+        data.slice(0, 1), // the first byte indexing the enum
+        data.slice(4, data.length), // the first byte indexing the enum
+      ])
+      return deserializeUnchecked(
+        MARINADE_BORSH_SCHEMA,
+        StakeState,
+        adjustedData,
+      )
+    })
+  }
+
+  async getStakeInfos(): Promise<StakeInfo[]> {
+    const stakeRecords = await this.getStakeRecords()
+    const stakeInfos = new Array<StakeInfo>()
+
+    const to_process = stakeRecords.length
+    let processed = 0
+    // rpc.get_multiple_accounts() has a max of 100 accounts
+    const BATCH_SIZE = 100
+    while (processed < to_process) {
```

Comment:
It would be nice if we used generators to abstract away the batching logic, but let's not worry about that for now.